### ISSUE TITLE
Unzip all ROMs on GitHub

### ISSUE THREAD

**Issue by vgturtle127** [schibo/1964js#7]
GitHub isn't meant for binary files or archives.

A better way would be to unzip them, then use a .gitattributes file to tell GitHub they are binary and to not modify them.

I can write a .gitattributes file if you'd like.

**Comment by schibo** [comment]
Right, we can probably host the files elsewhere but the rom files should remain zip compressed because the emulator uncompresses the archive at runtime.

**Comment by vgturtle127** [comment]
Oh, really? That's pretty cool! I didn't know it did that! That's fairly impressive. I didn't even notice in the browser anything had to load or anything! I tested using both Firefox developer edition and Chrome.

So you are using a JavaScript zip decompression library?

You could also use gh-pages too, if you wanted.

When I said you should unzip them, I was thinking it would be faster, but heck the speed is so good! I expected Firefox to have to really chug along if that was the case, but nope. All the demos run at full speed too. SO impressive.

There are some .v64 files in that directory as well, I'm assuming those aren't decompressed and just load as usual?

**Comment by schibo** [comment]
Yep, the uncompressed versions are just to test both compressed and uncompressed. The emu uses bitjs for unzipping. The main reason for using zip is to keep the download size smaller but It's great that it decompresses quickly too.

**Comment by vgturtle127** [comment]
Actually, if we are testing that, would you like it if I compresses the ROM images with various algorithms, encodings, and file formats to help better test the bitjs library? It says it has support for zip, tar, and rar. I'm assuming with tar support it might have gz and bzip support as well. I could also test encryption support as well. Just an idea. :smile:

**Comment by schibo** [comment]
That would be cool. There's certainly room for improvement. The reason the rom directory is named "unofficial_roms" is because I had to compress the ROMs myself. I vaguely recall that if the zip archive has more than one file in, it or if it has subdirectories, then it either doesn't uncompress properly or it doesn't find the file. The bitjs lib may have improved in the past few years. It might be worth updating the lib or seeing if there are better alternatives. Either that or I may have simply been using the unarchiver improperly.

**Comment by vgturtle127** [comment]
If I recall correctly, the bitjs source hasn't been touched since June of 2014. Since Google Code is shutting down soon, I went ahead and exported it to my GitHub to preserve it.

The bug you talk about actually happens with most archive libraries I have noticed. Only recently did this become an issue, mostly due to emulators and various other projects.

You have been using it properly, I think the bug is most likely due to the fact that the programmer didn't think people would need multiple files or directories or anything. Possibly he may have not even thought about it. Who knows.

Anyway, I found out that zip.js and rar.js are other libraries we could use. There is also JSZip, which is actually hosted on GitHub. Unfortunately, no tar support in any of them. I'm not sure why you would use tar for ROM images, but if you or somebody else had a good reason then we may have to just fork bitjs.

There have been attempts at porting all of 7-zip's code to JavaScript, but no complete port exists.

Also, maybe just calling the directory "roms" might be a good idea? I can see the underscore causing issues on some operating systems, Git clients, etc.

I also was wondering if you would like to add some more homebrew/public domain ROMs? THat help test the functionality of the emulator and stuff? Just an idea I had. I remember some ROM images I might have lying around that I use all the time for testing N64 emulators (since there are a lot of them and they almost always aren't perfect).

**Comment by schibo** [comment]
Yeah, before creating new archives for more demos, it would be nice to see if zip.js or JSZip can load official ROM archives properly. As I recall, none of the official ROMs tar or rar. If one of those zip libraries works the way we need it to, we could integrate it. It's probably better to rename the directory to "roms" after that works to avoid confusion with our unofficial roms. I can remove the underscore in the meantime if it's an issue.

**Comment by vgturtle127** [comment]
It wasn't an issue at the moment, but especially on some Linux distros I have noticed that when I have any special characters in my GitHub folder everything starts to explode. :laughing: 

I will see about the two other zip libraries and report back. :smile:
